Sources/Strand.swift

```diff
@@ -1,128 +1,92 @@
+*/
 
 #if os(Linux)
-import Glibc
+    import Glibc
 #else
-import Darwin.C
+    import Darwin.C
 #endif
 
-#if !swift(>=3.0)
-    typealias Error = ErrorType
-    typealias OpaquePointer = COpaquePointer
-#endif
+private class Box<T> {
+    let value: T
+    init(_ value: T) {
+        self.value = value
+    }
+}
 
 public enum StrandError: Error {
-    case threadCreationFailed
-    case threadCancellationFailed(Int)
-    case threadJoinFailed(Int)
+    case creationFailed(Int)
+    case cancellationFailed(Int)
+    case joinFailed(Int)
+    case detachFailed(Int)
 }
 
 public class Strand {
-    #if swift(>=3.0)
-        #if os(Linux)
-            private var pthread: pthread_t = 0
-        #else
-            private var pthread: pthread_t
-        #endif
-    #else
+    public typealias Closure = () -> Void
+
+    private var pthread: pthread_t
+
+    public init(_ closure: Closure) throws {
+        let box = Box(closure)
+        let holder = Unmanaged.passRetained(box)
+        let closurePointer = UnsafeMutablePointer<Void>(holder.toOpaque())
+
         #if os(Linux)
-            private var pthread: pthread_t = 0
+            var thread: pthread_t = 0
         #else
-            private var pthread: pthread_t = nil
-        #endif
-    #endif
-
-    public init(closure: () -> Void) throws {
-        let holder = Unmanaged.passRetained(StrandClosure(closure: closure))
-
-        #if swift(>=3.0)
-            let pointer = UnsafeMutablePointer<Void>(holder.toOpaque())
-            #if os(Linux)
-                guard pthread_create(&pthread, nil, runner, pointer) == 0 else {
-                    holder.release()
-                    throw StrandError.threadCreationFailed
-                }
-            #else
-                var pt: pthread_t?
-                guard pthread_create(&pt, nil, runner, pointer) == 0 && pt != nil else {
-                    holder.release()
-                    throw StrandError.threadCreationFailed
-                }
-                pthread = pt!
-            #endif
-        #else
-            let pointer = UnsafeMutablePointer<Void>(OpaquePointer(bitPattern: holder))
-            guard pthread_create(&pthread, nil, runner, pointer) == 0 else {
-                holder.release()
-                throw StrandError.threadCreationFailed
-            }
+            var thread: pthread_t?
         #endif
+
+        let result = pthread_create(&thread, nil, runner, closurePointer)
+        // back to optional so works either way (linux vs macos).
+        let inner: pthread_t? = thread
+
+        guard result == 0, let value = inner else {
+            holder.release()
+            throw StrandError.creationFailed(Int(result))
+        }
+        pthread = value
+    }
+
+    deinit {
+        pthread_detach(pthread)
```

Comment:
Since theres now a way to explicitly detach a thread we should have a flag to remember if the thread has been detached. During deinit, we shouldn't detach a second time if it has been done before. We should also add that guard in the detach method itself.